Dynamic colour picker changes on MSW.

Trylz opened the pull request:
This pull request adds two new events to the wxColourPickerCtrl.
wxEVT_COLOUR_SELECTED: Called when the dialog selected color is changed.
wxEVT_COLOUR_CANCELED: Called when the dialog is closed or the user clicks on CANCEL.

The implementation, which is MSW only has been built from the following answers:
https://social.msdn.microsoft.com/Forums/en-US/c5fcfd9f-6b27-4848-bb9d-94bec105eabd/get-the-current-clicked-color-from-choosecolor-dialog?forum=windowsgeneraldevelopmentissues

The selected color corresponds to the current color for both user clicks and the Api client.
A Call to SetColour on the wxColourPickerCtrl will also set the selected color to that value.
A Call to GetSelectedColour will always return the current "desired" color allowing to not care about the dialog state.
The feature can be a little bit confusing or is probably too specific. Don't hesitate if you have improvements.

Here a gif showing a use case:
![picker](https://user-images.githubusercontent.com/9217999/52181140-c42b0780-27bc-11e9-8ea6-bb204ef411f3.gif)

Commit:
Start work

Commit:
Implement new events

Commit:
Add new GetSelectedColour function

Commit:
Update colordlg.cpp

Commit:
Update clrpickerg.cpp

Commit:
Handle multiple dialogs case

Commit:
Update colordlg.cpp

Commit:
Remove reinterpret_cast to be consistent with rest of code

Commit:
use WM_NCDESTROY instead of WM_DESTROY

Commit:
Update colordlg.cpp

Cleanup before PR

Commit:
Put back original chooseColorStruct construct order

Commit:
wxASSERT_MSG instead of wxASSERT

Commit:
More robust selection allowing keyboard events.

Commit:
An attempt to fix the build

vadz reviewed (changes requested):
I had no idea about the hack with `COLOR_PROP`, thanks for finding out about it (I wonder how was it found originally). I do wonder if it's not something that risks exploding in our faces if a future Windows version changes this completely undocumented behaviour. It looks very unlikely, knowing that it has apparently been like this for 20+ years, but who knows...

So perhaps it would be safer to have a style explicitly requesting doing this subclassing? Or are we ready to take this risk? I'm not sure to be honest... @csomor If you're reading this, by chance, could such style be useful for wxMac too in order to allow the dialog to behave in the native (non-modal) way there when it's specified? Or do we need something more than this there anyhow?

Until we decide whether we need this flag or not, could you please address the comments below and also add

1. Documentation for the new events (see the files in `interface/wx` directory).
2. Demonstration of using them in the dialogs sample, to allow testing this easily.

? TIA!

vadz committed:
Fix src/msw/colordlg.cpp L130  typo

Co-Authored-By: Trylz <nebularender@gmail.com>

vadz committed:
Update src/msw/colordlg.cpp

Co-Authored-By: Trylz <nebularender@gmail.com>

vadz committed:
Use ProcessWindowEvent instead of GetEventHandler()->ProcessEvent

Co-Authored-By: Trylz <nebularender@gmail.com>

MaartenBent commented:
> Demonstration of using them in the dialogs sample, to allow testing this easily.

Adding it to the widgets sample is more appropriate, al already has a page with `wxColourPickerCtrl `. The dialogs sample only uses `wxColourDialog`.

vadz commented:
> 
> 
> > Demonstration of using them in the dialogs sample, to allow testing this easily.
> 
> Adding it to the widgets sample is more appropriate, it already has a page with `wxColourPickerCtrl `. The dialogs sample only uses `wxColourDialog`.

I meant to demonstrate it with `wxColourDialog` as this is where this functionality is primarily implemented, `wxColourPickerCtrl` just forwards it on, basically. Perhaps ideally we could have it in both...

csomor commented:
Hi, that's quite close to the native macOS color picker behavior, only there these would be a floating panel instead of a modal dialog, but I'd be fine with adding these two events but I think we'd need to maintain the dialog modal style on both platforms, so as I see it, we'd keep the mac version as a modal one.

Trylz commented on src/msw/colordlg.cpp:
```diff
@@ -61,6 +62,93 @@ wxIMPLEMENT_DYNAMIC_CLASS(wxColourDialog, wxDialog);
+#define COLORPROP (LPCTSTR) 0xA000L
+#define SUBCLASSING_PROP L"PROP_SUBCLASSING"
+
+struct COLORINFO
```
WIll fix it :)

vadz committed:
Update src/msw/colordlg.cpp L98 typo

Co-Authored-By: Trylz <nebularender@gmail.com>

vadz committed:
rename DialogSubclassingData

Co-Authored-By: Trylz <nebularender@gmail.com>

vadz committed:
Use static cast instead of c casting

Co-Authored-By: Trylz <nebularender@gmail.com>

vadz committed:
Update src/msw/colordlg.cpp L119 typo fix

Co-Authored-By: Trylz <nebularender@gmail.com>

vadz committed:
Use wxFAIL_MSG instead of wxASSERT_MSG

Co-Authored-By: Trylz <nebularender@gmail.com>

Trylz commented on src/msw/colordlg.cpp:
```diff
@@ -61,6 +62,93 @@ wxIMPLEMENT_DYNAMIC_CLASS(wxColourDialog, wxDialog);
 
+#define COLORBOXES 64
+#define COLORPROP (LPCTSTR) 0xA000L
+#define SUBCLASSING_PROP L"PROP_SUBCLASSING"
```
I will rename it for wxSUBCLASS_DATA. I was not sure for the multiple dialog case that is why I used props.
Thanks I will use a global variable to store the subclassing data.

Commit:
Apply review changes

- Remove selected color from clrpicker
- Remove colordlg  props usage
- Some typo fix and renaming

tomay3000 commented on src/msw/colordlg.cpp:
```diff
@@ -108,17 +108,21 @@ struct DialogSubclassingData
+    }
 
-    // Let the original procedure process the event first.
+    // Call the original procedure process the event first.
```
Wrong phrase sentence.

Trylz replied:
Fixed thanks :)

PBfordev commented on include/wx/clrpicker.h:
```diff
@@ -159,13 +161,15 @@ class WXDLLIMPEXP_CORE wxColourPickerCtrl : public wxPickerBase
 
 wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COLOURPICKER_CHANGED, wxColourPickerEvent );
+wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COLOUR_SELECTED, wxColourPickerEvent );
+wxDECLARE_EXPORTED_EVENT( WXDLLIMPEXP_CORE, wxEVT_COLOUR_CANCELED, wxColourPickerEvent );
```
I believe that wxWidgets use spelling with two Ls, i.e., "canceled" should be changed to "cancelled" everywhere.

Trylz replied:
Thanks i will fix it :)

Commit:
Rename wxEVT_COLOUR_CANCELED for wxEVT_COLOUR_CANCELLED

Commit:
Update documentation

Commit:
Update colour picker sample

Trylz commented:
I have updated the documentation and added a demonstration in the samples.
Should I implement the new style? I would call it  wxCLRP_SELECT_NOTIFICATIONS :).

Commit:
Remove redundant word

Commit:
Add " Currently only implemented on wxMSW" text

Trylz commented on include/wx/msw/colordlg.h:
```diff
@@ -32,6 +32,8 @@ class WXDLLIMPEXP_CORE wxColourDialog : public wxDialog
 
     wxColourData& GetColourData() { return m_colourData; }
 
+    void OnColorSelected(const wxColour& colour);
```
It can't be private since it is called from the hook procedure.

Commit:
Put pack old reinterpret_cast

Trylz commented:
Please can this be merged?  @vadz

PBfordev commented on src/common/clrpickercmn.cpp:
```diff
@@ -64,12 +67,19 @@ bool wxColourPickerCtrl::Create( wxWindow *parent, wxWindowID id,
     m_picker->Bind(wxEVT_COLOURPICKER_CHANGED,
             &wxColourPickerCtrl::OnColourChange, this);
 
+    m_picker->Bind(wxEVT_COLOUR_SELECTED,
```
Whitespace nitpick: The second-line indentation for Bind()ing the two new events does not match indentation above?

PBfordev commented on src/generic/clrpickerg.cpp:
```diff
@@ -86,15 +86,25 @@ void wxGenericColourButton::OnButtonClick(wxCommandEvent& WXUNUSED(ev))
-        GetEventHandler()->ProcessEvent(event);
-    }
+    const int res = dlg.ShowModal();
+    wxASSERT(res == wxID_OK || res == wxID_CANCEL);
```
Is the `wxASSERT()` needed here?

PBfordev commented on src/msw/colordlg.cpp:
```diff
@@ -76,11 +162,18 @@ wxColourDialogHookProc(HWND hwnd,
+        g_subclassingData.selectedColor = pCC->rgbResult;
+        g_subclassingData.colourDialog = dialog;
 
+        // Init the window title.
```
Are the two newly added comments for existing code (lines 171 and 176) necessary ? The code seems quite self-documenting...

vadz commented:
Sorry for the delay, but there were multiple things I was unhappy about in this PR. First of all, there seems to be no reason at all to conflate `wxColourDialog` event with `wxColourPickerCtrl` one, the functionality that you added can be useful when using the dialog directly too, so I had to decouple it from the picker.

Second, the implementation looked weird: I see no reason to subclass the dialog when we already have the hook procedure and I also don't see the point of forwarding the events from `wxGenericColourButton` in `wxColourPickerCtrl` rather than just generating the correct events directly.

I've created #1301 with my proposed replacement, so I'm closing this one as it's redundant now, but please test the new PR and let me know if you have any comments about it. TIA!
